---
layout : default
title : TnT Genome
type : overview
theme : genesAndTranscripts
---

<h2>Genome Browser based on TnT Board</h2>
{% include theme.html %}

<p><aside>See <a href={{ site.baseurl }}/examples/index.html>more examples</a></aside></p>

<h2><a name="introduction" href="#introduction">#</a>Introduction</h2>
<p><b>TnT Genome</b> is a javascript library for displaying genome based annotations. It has been built on top of the <a href="">TnT Board</a> library and shares with it the same phylosophy and design principles: a powerful API written with flexibility in mind that allows the creation of simple, dynamic and useful genome-based visualisations. The goal of TnT Genome is not to provide a full-featured genome browser but a simple tool to create specific and simple genome views. Its flexibility also allows the exploration of new ways of visualising genome-based annotations.</p>

<p>Like TnT Board, TnT Genome defines several elements to facilitate the creation of visualisations: a board that is a container for tracks, tracks that are independent units of data representation, data retrievers to get the data synchronously or asynchronously and visual features to display the available data. Check the <a href="">documentation for TnT Board</a> to get more details about these elements. What TnT Genome offers is new features for genes and transcripts, new data retrievers for ensembl data (using the Ensembl Rest API) and a layout for distributing genes and transcripts avoiding overlaps between elements. Check the <a href="">API documentation</a> to get more details about these new elements</p>


<h2><a name="example" href="#example">#</a>Example</h2>
<p>Examples are a fundamental part of TnT documentation. Check more in the examples page</p>

{% highlight javascript linenos %}
// Genome is a board. Location and axis tracks are included by default
var genome = tnt.board.genome().species("human").gene("brca2").width(950);

var gene_track = tnt.board.track()
    .height(200)
    .color("white")
    .display(tnt.board.track.feature.genome.gene()
        .color("#550055")
    )
    .data(tnt.board.track.data.genome.gene());

var sequence_track = tnt.board.track()
        .height(30)
        .background_color("white")
        .display(tnt.board.track.feature.genome.sequence())
        .data(tnt.board.track.data.genome.sequence()
            .limit(150)
        );
genome
    .zoom_in(100)
    .add_track(sequence_track)
    .add_track(gene_track);

genome(document.getElementById("mydiv"));
genome.start();
{% endhighlight %}

<p>The above example starts by defining a genome board using the human BRCA2 gene as its reference. The code also defines two tracks. The first one is a gene track and the second one is a sequence track. These tracks pulls data from the Ensembl REST API using the coordinates provided by the board (ie, BRCA2 coordinates). Once defined, the tracks are attached to the board. Once the board is configured the visualisation is ready to start.</p>


<h2><a name="installation" href="#installation">#</a>Installation</h2>
<p>TnT Genome can be linked directly from the github repo. It depends on the <a href="http://d3js.org">d3js</a> and this library needs to be linked before tnt.genome</p>

{% highlight html linenos %}
<link rel="stylesheet" href="http://emepyc.github.io/tnt/build/tnt.genome.css" type="text/css" />
<script src="http://d3js.org/d3.v3.min.js" charset="utf-8"></script>
<script src="http://emepyc.github.io/tnt/build/tnt.genome.js" charset="utf-8"></script>
{% endhighlight %}

<p>It is also possible to install the source code and build the library locally. This can be achieved either by using <b>npm</b> or <b>github</b>:</p>

<ul>
    <li>
        Using <b>npm</b>:
        {% highlight bash %}
npm install tnt.genome
        {% endhighlight %}
    </li>
    <li>
        Building from <b>Git</b>:
        {% highlight bash %}
git clone https://github.com/emepyc/tnt.genome
cd tnt.genome
npm install
npm build-browser
        {% endhighlight %}
    </li>
</ul>


<h2><a name="usage" href="#usage">#</a>Usage</h2>

<p><aside>Check out the <a href={{ site.baseurl }}/api/index.html>API</a></aside></p>

<p>Once the library has been installed and built, it can be used directly:</p>

{% highlight html linenos %}
<link rel="stylesheet" href="http://emepyc.github.io/tnt.genome/buid/tnt.genome.css" type="text/css" />
<script src="http://d3js.org/d3.v3.min.js" charset="utf-8"></script>
<script src="http://emepyc.github.io/tnt.genome/build/tnt.genome.min.js" charset="utf-8"></script>
{% endhighlight %}


<h2><a name="dependencies" href="#dependencies">#</a>Dependencies</h2>
<p>TnT Genome has been built on top of the excellent d3js library, being this its main dependency. Experience with d3js is only needed to develop extensions (like new visual features). D3js is not included in the library and has to be included in your web page <i>before</i> including TnT Genome.</p>

<h2><a name="browsers" href="#browsers">#</a>Browser support</h2>
<p>TnT works on HTML5, CSS3 and SVG-compliant browser, which generally means any "modern browser". The TnT library and themes are tested regularly against Chrome (Chromium) and major changes are regularly tested in recent versions of Firefox, Safari (Webkit) and Opera. It has been also tested in IE versions 9+. If you find any problem, please <a href="https://github.com/emepyc/tnt/issues">let me know</a>.</p>

<p>The TnT library depends on the <a href="http://d3js.org">D3.js</a> library. Check its <a href="https://github.com/mbostock/d3/wiki#browser-support">browser's compatibility note</a> for more information.</p>
